[rocFFT] Fixes for Hermitian symmetrizer

eng-flavio-teixeira opened the pull request:
## Motivation

 - Fixes issue with length < 3 inputs in the device-side Hermitian symmetrizer.

## Technical Details

 - Fixed griddim generation for length < 3.

## Test Plan

 - Added compare_cpu_gpu_symmetrizers unit test to compare symmetrizer output for several input parameters. 
 - Added adhoc accuracy tests that exercise the problematic symmetrizer input parameters.

## Test Result

 - Newly created tests confirm that both host- and device-sided symmetrizers produce same results, and that length < 3 works as intended.

## Submission Checklist

- [ ] Look over the contributing guidelines at https://github.com/ROCm/ROCm/blob/develop/CONTRIBUTING.md#pull-requests.

eng-flavio-teixeira committed:
- Decouple Hermitian symmetrizer from USE_HIPRAND.

eng-flavio-teixeira committed:
Add ability to choose between cpu or gpu Hermitian symmetrizer in compute_input()

eng-flavio-teixeira committed:
Fix generate_hermitian_gridDim for length < 3

eng-flavio-teixeira committed:
Add unit test

regan-amd commented on projects/rocfft/shared/data_gen_device.h:
```diff
@@ -65,6 +65,65 @@ __device__ static void set_imag_zero(const size_t pos, Tfloat* xreal, Tfloat* xi
+    if(length.size() > 1 && length[0] == 0)
+        throw std::runtime_error("Invalid zero length for impose_hermitian_symmetry");
+
+    if(length.size() > 2 && length[1] == 0)
```
```suggestion
    if(!length.empty() && std::count(length.begin(), length.end()-1, static_cast<size_t>(0)) > 0)
```
\+ while we're at it, shall we validate that `batch > 0` as well?

amd-jnovotny commented:
Are these fixes significant enough to note in the changelog?

regan-amd commented on projects/rocfft/shared/fft_params.h:
```diff
@@ -4013,7 +4175,8 @@ void init_local_input(int                                       comm_rank,
                                           contiguous_stride,
-                                          contiguous_dist);
+                                          contiguous_dist,
+                                          true);
```
The added constant-valued argument should be `Tbuff`-dependent in my understanding. Could we declare `set_input` with consistent default values and rely on those instead?
Note: this suggestion is irrelevant if it's decided to leave the `set_input` functions unchanged in the end.

eng-flavio-teixeira committed:
- Add adhoc accuracy test cases that exercise small lengths in the symmetrizer.
 - CHANGELOG.

eng-flavio-teixeira commented:
> Are these fixes significant enough to note in the changelog?

I believe so, thanks for pointing that out. I've added the entry in changelog.

amd-jnovotny commented:
@eng-flavio-teixeira Thanks. Changelog updates look good.

regan-amd commented on projects/rocfft/clients/tests/hermitian_test.cpp:
```diff
@@ -355,3 +355,104 @@ TEST(rocfft_UnitTest, gpu_symmetrizer)
+#ifdef USE_HIPRAND
+            for(size_t i = 0; i < nbuffer; ++i)
+            {
+                ASSERT_TRUE(gpu_symm_gpubuf[i].alloc(p.ibuffer_sizes()[i]) == hipSuccess);
```
Do we want to fail here (and after other similar runtime-triggered errors) even if `--skip_runtime_fails` is used, _e.g._, for default test?

regan-amd commented on projects/rocfft/shared/fft_params.h:
```diff
@@ -366,7 +430,8 @@ inline void set_input(std::vector<hostbuf>&      input,
                       const Tint1                field_contig_stride,
-                      const size_t               field_contig_dist)
+                      const size_t               field_contig_dist,
+                      const bool                 use_gpu_symmetrizer)
```
```suggestion
                      const bool                 use_gpu_symmetrizer = false)
```
to keep its default behavior unchanged (unless it's decided to leave this unchanged in the end).

regan-amd commented on projects/rocfft/shared/fft_params.h:
```diff
@@ -241,7 +241,8 @@ inline void set_input(std::vector<gpubuf>&       input,
                       const Tint1&               field_contig_stride,
-                      const size_t               field_contig_dist)
+                      const size_t               field_contig_dist,
+                      const bool                 use_gpu_symmetrizer)
```
```suggestion
                      const bool                 use_gpu_symmetrizer = true)
```
to keep its default behavior unchanged (unless it's decided to leave this unchanged in the end).

regan-amd commented on projects/rocfft/clients/tests/hermitian_test.cpp:
```diff
@@ -355,3 +355,104 @@ TEST(rocfft_UnitTest, gpu_symmetrizer)
+            p.precision      = fft_precision_double;
+            p.transform_type = fft_transform_type_real_inverse;
+            p.itype          = itype;
+            p.placement      = fft_placement_notinplace;
```
I'd suggest to use a non-default (non-unit) batch size to cover the corresponding logic in the symmetrizing kernels/routines as well.

regan-amd commented on projects/rocfft/shared/fft_params.h:
```diff
@@ -1996,7 +2155,7 @@ class fft_params
     // Given a data type and dimensions, fill the buffer, imposing Hermitian symmetry if necessary.
     template <typename Tbuff>
-    inline void compute_input(std::vector<Tbuff>& input)
+    inline void compute_input(std::vector<Tbuff>& input, const bool use_gpu_symmetrizer = true)
```
```suggestion
    inline void compute_input(std::vector<Tbuff>& input, const bool use_gpu_symmetrizer = default_symmetrizer<Tbuff>::use_gpu)
```
Unless I'm mistaken, we currently use the host (resp. gpu) symmetrizer for host (resp. gpu) input buffers so we'd need the default value to depend on the template specialization to keep the same behavior.
This suggestion would achieve that with a `default_symmetrizer` struct defined as
```
template <typename Tbuff>
struct default_symmetrizer;

template <typename T>
struct default_symmetrizer<hostbuf_t<T>>
{
    static constexpr bool use_gpu = false;
};

template <typename T>
struct default_symmetrizer<gpubuf_t<T>>
{
    static constexpr bool use_gpu = true;
};
```
Note 1: the same type_trait struct could be used for declaring default values of the `set_input` variants as well and ensure consistency between the default values for `compute_input` and the corresponding `set_input` it intends to use.
Note 2: this suggestion is irrelevant if it's decided to leave this function unchanged in the end.

regan-amd commented on projects/rocfft/clients/tests/hermitian_test.cpp:
```diff
@@ -355,3 +355,104 @@ TEST(rocfft_UnitTest, gpu_symmetrizer)
+            }
+
+            p.igen = fft_input_random_generator_device;
+            p.compute_input(gpu_symm_gpubuf, true);
+            p.compute_input(cpu_symm_gpubuf, false);
```
This usage of `compute_input` in this test basically hides the actual targets (_i.e._, the various `impose_hermitian_symmetry_*` functions) and it is the only motivation that I can find for making it (and the `set_input` functions) accept an extra `bool` argument. These added arguments seem rather invasive, and an inadequate usage thereof would trigger numerous undesirable copies between the host and the device in tests. Also, unless I'm mistaken, this test basically relies on the generation of the random input sequence being stable for similar data layouts, which seems like an otherwise breakable design choice to me.

Wouldn't it be preferrable (and/or more future-proof) to invoke the appropriate `generate_random_interleaved_data` variant _once_ (in this case on a gpu buffer), _copy_ the content over to the second buffer (in this case a host buffer), invoke the corresponding symmetrizing routine on either buffer (host routine on host buffer, device routine on gpu buffer), and eventually compare both outputs (after bringing the content of the gpu buffer back to the host)?
In my opinion, that would trigger fewer copies, leave the `compute_input` and `set_input` functions unchanged (leaving their default/only behavior unchanged), and it would also make this test more self-explanatory.

eng-flavio-teixeira replied:
Yeah, the core issue here is that the symmetrizer and data generators are tightly coupled. My first attempt was to decouple them, but this ends up being way more work than it should for this test. I have pushed a new commit that should simplify things.

regan-amd commented on projects/rocfft/clients/tests/hermitian_test.cpp:
```diff
@@ -355,3 +355,104 @@ TEST(rocfft_UnitTest, gpu_symmetrizer)
+    const double                           double_linf_cutoff = 1e-15;
+    std::vector<std::pair<size_t, size_t>> linf_failures;
+
+    const auto array_types
+        = {fft_array_type_hermitian_interleaved, fft_array_type_hermitian_planar};
```
If achievable, being capable of cross-checking between the two array types (_e.g._, symmetrizing a planar array on the host and symmetrizing an interleaved array holding the same complex values on the device) would make all test targets even more robust in my opinion.

eng-flavio-teixeira committed:
- Simplify hermitian test.
 - Revert change fft_params changes.
 - Add batch to hermitian tests.

eng-flavio-teixeira committed:
- Add igen check in set_input

regan-amd commented on projects/rocfft/shared/fft_params.h:
```diff
@@ -243,6 +243,9 @@ inline void set_input(std::vector<gpubuf>&       input,
                       const Tint1&               field_contig_stride,
                       const size_t               field_contig_dist)
 {
+    if(igen == fft_input_generator_host || igen == fft_input_random_generator_host)
+        throw std::runtime_error("Host input generation is not available for gpu buffers");
+
```
Thanks for that!

regan-amd commented on projects/rocfft/clients/tests/hermitian_test.cpp:
```diff
@@ -355,3 +355,99 @@ TEST(rocfft_UnitTest, gpu_symmetrizer)
+        }
+    }
+#else
+    GTEST_SKIP();
```
```suggestion
    GTEST_SKIP() << "this test currently requires hiprand to be used";
```

eng-flavio-teixeira committed:
- Handle potential runtime failures in new test.

eng-flavio-teixeira committed:
- Add batch size validation.
  - Simplify length array check.

regan-amd commented on projects/rocfft/clients/tests/hermitian_test.cpp:
```diff
@@ -355,3 +355,127 @@ TEST(rocfft_UnitTest, gpu_symmetrizer)
+                    cpu_symm_hostbuf[i].alloc(p.ibuffer_sizes()[i]);
+                }
+
+                p.igen = fft_input_generator_device;
+                p.compute_input(gpu_symm_gpubuf);
+
+                p.igen = fft_input_generator_host;
+                p.compute_input(cpu_symm_hostbuf);
```
Side comment: there is no more randomization in the input generation yet the test does require hiprand to be used: it's a little strange. I understand this is not of your own making, but I think we could/should plan a follow-up task to address that oddity (it doesn't seem overwhelmingly hard).

eng-flavio-teixeira replied:
Agree, it is a little odd that we need USE_HIPHAND for a deterministic input.

regan-amd commented on projects/rocfft/clients/tests/hermitian_test.cpp:
```diff
@@ -355,3 +355,127 @@ TEST(rocfft_UnitTest, gpu_symmetrizer)
+
+                    cpu_symm_hostbuf[i].alloc(p.ibuffer_sizes()[i]);
+                }
+
```
```suggestion
// This test does rely on the (deterministic) host and device input generators producing the same data sequences
```

regan-amd reviewed (approved):
Thank you! If you could have the changes to `data_gen_device.h` applied to the hipfft folder as well, that would be appreciated.

evetsso commented on projects/rocfft/clients/tests/accuracy_test_adhoc.cpp:
```diff
@@ -349,6 +349,8 @@ const auto adhoc_tokens = {
     "real_forward_len_378_42_single_ip_batch_66000_istride_44_1_R_ostride_22_1_HI_idist_16632_odist_8316_ioffset_0_0_ooffset_0_0",
     "real_forward_len_527_25_single_ip_batch_67500_istride_26_1_R_ostride_13_1_HI_idist_13702_odist_6851_ioffset_0_0_ooffset_0_0",
     "real_forward_len_630_38_single_ip_batch_65540_istride_40_1_R_ostride_20_1_HI_idist_25200_odist_12600_ioffset_0_0_ooffset_0_0",
+    "real_inverse_len_2_5_7_double_ip_batch_1_istride_20_4_1_HI_ostride_40_8_1_R_idist_40_odist_80_ioffset_0_0_ooffset_0_0",
```
Ah, we don't currently allow inverse tokens in adhoc tests since they will do round-trips, and the round-trip code is only expecting forward tokens so that it can inverse-ify them.

Workaround is to make this a forward token that will become this inverse token when the round-trip code gets to it.

regan-amd replied:
But the symmetrizers wouldn't be hit if we did that unless I'm mistaken. Could we add a `&& params.is_forward()` to the definition of [`round_trip`](https://github.com/ROCm/rocm-libraries/blob/9c1e6f953cf131bd036f2c6fa405893f30982664/projects/rocfft/clients/tests/rocfft_accuracy_test.cpp#L83) instead?

evetsso replied:
Ah right, yeah it would make more sense to just not attempt the round-trip in that case.

eng-flavio-teixeira replied:
Yeah, good catch. I've added a field in fft_params only for this purpose. I'm worried that relying on is_forward() for the round trip check may affect other tests.

regan-amd replied:
For my own sanity: do you have a specific example/instance of how adding a check on `is_forward()` could affect other tests? It is in fact a legitimate condition (_i.e._, a requirement) for enabling roundtrip verifications in the accuracy test as it is currently implemented. In my personal opinion, it is actually _missing_ in the definition of [round_trip](https://github.com/ROCm/rocm-libraries/blob/9c1e6f953cf131bd036f2c6fa405893f30982664/projects/rocfft/clients/tests/rocfft_accuracy_test.cpp#L83), and addressing that seems easier and preferrable to adding a public, test-specific member to all operation-defining structs inheriting non-privately from `fft_params` (it would also align with hipfft's [test logic](https://github.com/ROCm/rocm-libraries/blob/4e0155037c8142300f44b2da942035d6d1d7ab0c/projects/hipfft/clients/tests/hipfft_accuracy_test.cpp#L111)).

eng-flavio-teixeira replied:
Right, there are no inverse test cases except the ones I've just added here. I've added the is_forward check.

eng-flavio-teixeira committed:
- Add data_gen_device changes to hipfft
 - Add comments to hermitian_test
 - Add fft_params field to check round trip in accuracy tests.
 - Add further token suite to exercise the Hermitian symmetrizer in accuracy tests.

eng-flavio-teixeira committed:
Merge branch 'develop' into hermitian_symmetry_tests_and_fixes

eng-flavio-teixeira committed:
- Remove check_round_trip from fft_params.

eng-flavio-teixeira committed:
Merge branch 'hermitian_symmetry_tests_and_fixes' of https://github.com/eng-flavio-teixeira/rocm-libraries into hermitian_symmetry_tests_and_fixes

eng-flavio-teixeira committed:
Merge branch 'develop' into hermitian_symmetry_tests_and_fixes

eng-flavio-teixeira committed:
Merge branch 'develop' into hermitian_symmetry_tests_and_fixes

evetsso reviewed (approved):
You'll need to move the changelog entry to the new section, now that #2018 is merged.

evetsso committed:
Merge remote-tracking branch 'upstream/develop' into hermitian_symmetry_tests_and_fixes

evetsso committed:
move changelog

evetsso committed:
only change forward to inverse if the token really is forward

evetsso committed:
real in-place cannot use hermitian planar

evetsso committed:
Merge remote-tracking branch 'upstream/develop' into hermitian_symmetry_tests_and_fixes
